lib/ansible/playbook/play.py

```diff
@@ -306,29 +306,35 @@ def compile(self):
 
             b = Block(play=self)
             b.block = self.pre_tasks or [noop_task]
-            b.always = [flush_block]
+            if self.pre_tasks:
+                b.always = [flush_block]
```

Comment:
I think we should just do (and similar with `tasks+roles` and `post_tasks` below):
```python
if self.pre_tasks:
    b = Block(play=self)
    b.block = self.pre_tasks
    b.always = [flush_block]
    block_list.append(b)
```

otherwise we would just append a block with a noop task which would needlessly run that noop task.